File: app/src/main/java/mx/edu/upqroo/kristenandroid/api/kristen/KristenApiConverter.java
package mx.edu.upqroo.kristenandroid.api.kristen;

import android.app.Application;
import android.os.AsyncTask;

import java.util.ArrayList;
import java.util.List;

import mx.edu.upqroo.kristenandroid.api.kristen.containers.Contenido;
import mx.edu.upqroo.kristenandroid.api.kristen.containers.Contenido_;
import mx.edu.upqroo.kristenandroid.api.kristen.containers.PublicacionContenido;
import mx.edu.upqroo.kristenandroid.data.database.entities.Contact;
import mx.edu.upqroo.kristenandroid.data.models.Content;
import mx.edu.upqroo.kristenandroid.data.models.ContentGallery;
import mx.edu.upqroo.kristenandroid.data.models.ContentImage;
import mx.edu.upqroo.kristenandroid.data.models.ContentLink;
import mx.edu.upqroo.kristenandroid.data.models.ContentList;
import mx.edu.upqroo.kristenandroid.data.models.ContentTitle;
import mx.edu.upqroo.kristenandroid.data.models.ContentVideo;
import mx.edu.upqroo.kristenandroid.data.models.NewsDetail;
import mx.edu.upqroo.kristenandroid.data.repositories.ContactRepository;

/**
 * <h1>KristenApiConverter</h1>
 * Class that converts the entities of the service into entities that are the views in the
 * application.
 */
class KristenApiConverter {

    static NewsDetail PublicacionToNewsDetail(PublicacionContenido post) {
        return new NewsDetail(post.getIdPublicaciones(),
                post.getUrl(),
                post.getTitulo(),
                post.getDescripcion(),
                post.getPortada(),
                post.getCategorias(),
                post.getFecha(),
                post.getIdTiposPublicacion(),
                post.getAutor(),
                publicationContentToNewsContent(post.getContenidos()));
    }

    private static List<Content> publicationContentToNewsContent(List<Contenido> contenidos) {
        List<Content> contents = new ArrayList<>();
        for (Contenido c : contenidos) {
            Contenido_ contenido = c.getContenido();
            if (c.getIdTipoContenidos().equals(1)) {
                contents.add(new Content(contenido.getTexto()));
            } else if (c.getIdTipoContenidos().equals(2)) {
                contents.add(new ContentImage(contenido.getAlt(),
                        contenido.getSrc()));
            } else if (c.getIdTipoContenidos().equals(4)) {
                contents.add(new ContentLink(contenido.getTexto(), contenido.getUrl()));
            } else if (c.getIdTipoContenidos().equals(5)) {
                contents.add(new ContentGallery(contenido.getImagenes()));
            } else if (c.getIdTipoContenidos().equals(7)) {
                contents.add(new ContentVideo(contenido.getId(), contenido.getServidor()));
            } else if (c.getIdTipoContenidos().equals(8)) {
                contents.add(new ContentList(contenido.getTitulo(), contenido.getOrdenada(),
                        contenido.getElementos()));
            } else if (c.getIdTipoContenidos().equals(9)) {
                contents.add(new ContentTitle(contenido.getTexto()));
            }
        }
        return contents;
    }

    /***
     * This method will receive a list of contacts to inset only if they are different from
     * the ones that are already in the database, in case that there is no contact in the databse
     * then they will be inserted without comparing anything.
     * @param contacts contacts from service
     * @param application application to get the repository instance
     */
    static void insertContacts(List<Contact> contacts, Application application) {
        ContactRepository repo = ContactRepository.getInstance(application);
        AsyncTask.execute(() -> {
            for (Contact contact: contacts) {
                repo.upsert(contact);
            }
        });
    }
}

File: app/src/main/java/mx/edu/upqroo/kristenandroid/api/kristen/KristenApiConverter.kt
package mx.edu.upqroo.kristenandroid.api.kristen

import android.app.Application
import android.os.AsyncTask

import java.util.ArrayList

import mx.edu.upqroo.kristenandroid.api.kristen.containers.Contenido
import mx.edu.upqroo.kristenandroid.api.kristen.containers.PublicacionContenido
import mx.edu.upqroo.kristenandroid.data.database.entities.Contact
import mx.edu.upqroo.kristenandroid.data.models.Content
import mx.edu.upqroo.kristenandroid.data.models.ContentGallery
import mx.edu.upqroo.kristenandroid.data.models.ContentImage
import mx.edu.upqroo.kristenandroid.data.models.ContentLink
import mx.edu.upqroo.kristenandroid.data.models.ContentList
import mx.edu.upqroo.kristenandroid.data.models.ContentTitle
import mx.edu.upqroo.kristenandroid.data.models.ContentVideo
import mx.edu.upqroo.kristenandroid.data.models.NewsDetail
import mx.edu.upqroo.kristenandroid.data.repositories.ContactRepository

/**
 * <h1>KristenApiConverter</h1>
 * Class that converts the entities of the service into entities that are the views in the
 * application.
 */
internal object KristenApiConverter {

    fun publicacionToNewsDetail(post: PublicacionContenido): NewsDetail {
        return NewsDetail(post.idPublicaciones,
                post.url,
                post.titulo,
                post.descripcion,
                post.portada,
                post.categorias,
                post.fecha,
                post.idTiposPublicacion!!,
                post.autor,
                publicationContentToNewsContent(post.contenidos!!))
    }

    private fun publicationContentToNewsContent(contenidos: List<Contenido>): List<Content> {
        val contents = ArrayList<Content>()
        for (c in contenidos) {
            val contend = c.contenido
            when {
                c.idTipoContenidos == 1 -> contents.add(Content(contend!!.texto!!))
                c.idTipoContenidos == 2 ->
                    contents.add(ContentImage(contend!!.alt!!, contend.src!!))
                c.idTipoContenidos == 4 ->
                    contents.add(ContentLink(contend!!.texto!!, contend.url!!))
                c.idTipoContenidos == 5 ->
                    contents.add(ContentGallery(contend!!.imagenes!! as MutableList<String>))
                c.idTipoContenidos == 7 ->
                    contents.add(ContentVideo(contend!!.id!!, contend.servidor!!))
                c.idTipoContenidos == 8 ->
                    contents.add(ContentList(contend!!.titulo!!, contend.ordenada!!,
                        contend.elementos!! as MutableList<String>))
                c.idTipoContenidos == 9 -> contents.add(ContentTitle(contend!!.texto!!))
            }
        }
        return contents
    }

    /***
     * This method will receive a list of contacts to inset only if they are different from
     * the ones that are already in the database, in case that there is no contact in the databse
     * then they will be inserted without comparing anything.
     * @param contacts contacts from service
     * @param application application to get the repository instance
     */
    fun insertContacts(contacts: List<Contact>, application: Application) {
        val repo = ContactRepository.getInstance(application)
        AsyncTask.execute {
            for (contact in contacts) {
                repo.upsert(contact)
            }
        }
    }
}
